[17기 김채현] PostingLikeView 작성

JELLO-Kim opened the pull request:
>wecode PR message template 및 체크 리스트 입니다. 
아래 사항들을 전부 작성/체크 하시고 PR 해주세요!

## 수정 사항 간략한 한줄 요약
(여기에 수정 사항에 대한 간략한 한줄 요약/제목 작성해 주세요.)
- 리뷰 반영, 좋아요와 스크랩 구현추가
- 비회원과 회원의 메인페이지 출력 다르게함
		 
## 수정 사항들 자세한 내용
(여기에 수정 사항에 대한 자세한 내용을 작성해 주세요.)
-  PostingLikeView에서 prefetch_related 사용
- 불필요한 변수 제거
- get메소드에 `@login_decorator`를 달아주어 토큰에 담긴 유저가 어느 게시물을 '좋아요' 하거나 '스크랩' 하면 유저id와 게시물id가 db의 해당하는 테이블에 저장 => 이후 get으로 화면 새로고침할때 토큰의 유저가 '좋아요'하거나 '스크랩'한 게시물에는 색이 채워진 아이콘으로 표시되도록 value값 전달 **(True / False)**
- 회원이 메인페이지(게시물)에 접근했을 경우 그 회원이 '좋아요' '스크랩'한 것은 채워진 이모티콘 보이게 함

## 기타 질문 및 특이 사항
(궁금한 사항들, 하면서 느낀 점들 공유해주실 것이 있으면 자유롭게 작성해 주세요.)
- 소헌님! views.py 60번째 줄에
```python
"like_status"  : True if posting.postinglike_set.filter(user=user) else False
```
로 작성한 부분이요.
토큰의 유저와 반복문도는 posting이 '좋아요' 관계에 있으면 **True**, 없으면 **False**를 반환해줘서 프론트님한테 드리면 프론트에서 `True일때 채워진 하트`, `False일땐 빈 하트`를 화면에 띄우게 하시기로 했습니다! 저런 방법도 괜찮나요?!
=> 통신으로 작동 확인했습니다!! scrap 도 같은 로직으로 작성했습니다. PostingScrapView는 준하님 담당이셔서 준하님이 올리실 꺼에요.

- `prefetch_related('comment')`를 해도 코멘트들 중 **첫번째**를 찾는 과정에서 계속 db에 접근하네요..ㅠㅡㅡㅜ... 이거랑 annotate로 만든것들도 prefetch 해주고싶지만.. 쉽지않네요......

## 체크 리스트 (아래 사항들이 전부 체크되어야만 merge가 됩니다!)
- [X] 필요한 test들을 완료하였고 기능이 제대로 실행되는지 확인 하였습니다.
- [X] Wecode의 코드 스타일 가이드에 맞추어 코드를 작성 하였습니다.
- [X] 제가 의도한 파일들과 수정 사항들만 커밋이 된 것을 확인 하였습니다.
- [X] 본 수정 사항들을 팀원들과 사전에 상의하였고 팀원들 모두 해당 PR에 대하여 알고 있습니다.

JELLO-Kim committed:
Add : PostingLikeView
- db에 값이 없을 경우 POST 결과 instance 생성(create)
- db에 해당 유저&게시글 조합에 대한 instance가 있을 경우 삭제(delete)

JELLO-Kim committed:
Add : PostingLikeView url path
- PostingLikeView에 연결되는 url 설정

JELLO-Kim committed:
Modify : order 조건에 영어이름 추가
- 프론트와 협의 후 URI에 띄워질 영어이름 추가

JELLO-Kim committed:
Modify : 로그인 유저의 게시물 좋아요 정보 추가
- 로그인 유저가 좋아요 상태인 게시물은 True를, 로그인 유저와 좋아요 관계가 없는 게시물에는 False를 반환한다
- True일 경우 채워진 하트, False일 경우 빈 하트를 프론트에서 구현한다.

soheon-lee commented on posting/views.py:
```diff
@@ -63,14 +69,32 @@ def get(self, request):
+    @login_decorator
+    def post(self, request):
+        try:
+            user        = request.user
+            data        = json.loads(request.body)
+            size_id     = PostingSize.objects.get(id=data['size']).id
+            housing_id  = PostingHousing.objects.get(id=data['housing']).id
+            style_id    = PostingStyle.objects.get(id=data['style']).id
+            space_id    = PostingSpace.objects.get(id=data['space']).id
+            image_url   = data['card_image']
```
`Class.objects.get(id=____)` 라는 부분을 보아, JSON body에서 이미 size나 housing, style, space의 `id`를 받아오고 있으신데, 이를 이용해서 다시 `PostingSize, PostingHousing, PostingStyle, PostingSpace`의 객체의 `id`를 받아오시는 이유가 따로 있을까요?

chaehyun-dev replied:
아뇨 이유 없습니다.. 수정하겠습니다!

soheon-lee commented on posting/views.py:
```diff
@@ -15,8 +18,10 @@
+    @login_decorator
     def get(self, request):
-        postings        = Posting.objects.all()
+        user            = request.user
+        postings        = Posting.objects.prefetch_related('comment', 'postinglike_set').select_related('user').all()
         order_request   = request.GET.get('order', 'recent')
```
역참조 하는 대상들은 `prefetch_related`, 정참조 하는 대상들은 `select_related` good good!
쿼리 수 많이 줄어들었겠네요 !

chaehyun-dev replied:
마지막 난관이 남았습니다... annotate랑 인덱싱 상황에서 하기..

soheon-lee commented on posting/views.py:
```diff
@@ -107,3 +131,20 @@ def get(self, request):
+        user = request.user
+        data = json.loads(request.body)
+        posting_id = data['posting_id']
+
+        posting = Posting.objects.get(id=posting_id)
+
```
여기도 PostingLike를 한 번 캐싱하시면 좋을 것 같아요 :)

chaehyun-dev replied:
넹~~~ 캐싱 너무좋아여!

soheon-lee commented on posting/views.py:
```diff
@@ -107,3 +131,20 @@ def get(self, request):
+        posting = Posting.objects.get(id=posting_id)
+
+        if posting.like_user.filter(id=user.id).exists():
+            posting = PostingLike.objects.filter(user_id=user.id, posting_id=posting_id)
+            posting.delete()
```
`posting`이란 변수 따로 선언 하시지 않고 바로 PostingLike 객체 delete하셔도 좋을 것 같네요 :)
가독성을 해치지 않는 선에서는,꼭 필요하지 않은 변수는 선언하지 않으셔도 됩니다 👍

chaehyun-dev replied:
넹~~~!!

soheon-lee commented on posting/views.py:
```diff
@@ -107,3 +131,20 @@ def get(self, request):
+        posting = Posting.objects.get(id=posting_id)
+
+        if posting.like_user.filter(id=user.id).exists():
+            posting = PostingLike.objects.filter(user_id=user.id, posting_id=posting_id)
```
posting.like_user.filter(id=user.id)로 줄이셔도 좋겠습니다 :) 위에 if문 처럼요!

chaehyun-dev replied:
넵!!! :)

soheon-lee commented on posting/views.py:
```diff
@@ -107,3 +131,20 @@ def get(self, request):
+        user = request.user
+        data = json.loads(request.body)
+        posting_id = data['posting_id']
+
+        posting = Posting.objects.get(id=posting_id)
+
+        if posting.like_user.filter(id=user.id).exists():
+            posting = PostingLike.objects.filter(user_id=user.id, posting_id=posting_id)
+            posting.delete()
+            return JsonResponse({'message' : 'SUCCESS'}, status=204)
+
+        posting.like_user.add(User.objects.get(id=user.id))
+        return JsonResponse({'message' : 'SUCCESS'}, status=201)
```
good good! 깔끔하게 정리 잘 하셨습니다 채현님. 아주... 마음에 드는... 깔끔함...

chaehyun-dev replied:
예~!~~!~!~!~!~!~!~!~!~!!~ 저도 하면서 너무 맘에 들었어요~!!! 감사합니닿ㅎ!!!!!

JELLO-Kim committed:
Modify : 리뷰사항 반영
- prefetch_related 추가
- 필요없는 변수 제거

JELLO-Kim commented:
소헌님! views.py 60번째 줄에
```python
"like_status"  : True if posting.postinglike_set.filter(user_id=user.id) else False
```
로 작성한 부분이요.
토큰의 유저와 반복문도는 posting이 '좋아요' 관계에 있으면 True, 없으면 False를 반환해줘서 프론트님한테 드리면 프론트에서 True일때 채워진 하트, False일땐 빈 하트를 화면에 띄우게 하시기로 했습니다! 저런 방법도 괜찮나요?!

JELLO-Kim committed:
Modify : 필요없는 삼항연산자 제거
- commnet, like, scrap 갯수 셀때 없으면 원래 0으로 들어가는것 확인 => if~ else 0 제거

JELLO-Kim committed:
Modify : like_status, scrap_status 정보 추가
- 토큰에 담긴 유저가 해당 게시물을 '좋아요'인지 '스크랩'인지 상태 보내줌

JELLO-Kim committed:
Merge branch 'main' of https://github.com/wecode-bootcamp-korea/17-1st-SweetHome-backend into feature/postingview

JELLO-Kim committed:
Modify : posting get에서 login_decorator 삭제
- 비회원도 메인페이지 볼수 있도록 수정

soheon-lee commented:
LGTM!

soheon-lee commented:
user 없을 때 정상 동작하는지 한 번 확인해주세요~

JELLO-Kim committed:
Modify : login_decorator 추가 / 비회원도 가능하도록
- 메인화면은 비회원도 조회할수 있도록 함
- 비회원일 경우 request.user에 None이 담기도록 함

JELLO-Kim committed:
Modify : 비회원에게 None 부여
- 토큰이 있을 경우 회원 반환
- 토큰이 없을 경우 error 대신 None 반환

JELLO-Kim committed:
Modify : decorator 추가
- 메인뷰 회원/비회원 구분용 데코레이터 : non_user_accept_decorator
- 오직 회원만 활동할 수 있는 데코레이터 : login_decorator

JELLO-Kim committed:
Modify : change url path
- 전체 경로에 posting 추가

JELLO-Kim committed:
Merge branch 'main' into feature/postingview
Fix : confilct

soheon-lee commented on posting/views.py:
```diff
@@ -44,33 +49,53 @@ def get(self, request):
+            image_url   = data['card_image']
+            content     = data['card_content']
+
+            Posting.objects.create(user_id=user.id, image_url=image_url, content=content, size_id=size_id, housing_id=housing_id, style_id=style_id, space_id=space_id)
```
```suggestion
            Posting.objects.create(
                user_id=user.id,
                image_url=image_url,
                content=content,
                size_id=size_id,
                housing_id=housing_id,
                style_id=style_id,
                space_id=space_id
            )
```
`=`도 정렬해주세요~

soheon-lee commented on user/utils.py:
```diff
@@ -1,9 +1,27 @@
+from my_settings    import SECRET_KEY, ALGORITHM
+from user.models    import User
+
+def non_user_accept_decorator(func):
```
이름 good!

JELLO-Kim committed:
Modify : Convention 수정

JELLO-Kim commented:
- `prefetch_related('comment')`를 해도 코멘트들 중 **첫번째**를 찾는 과정에서 계속 db에 접근하네요..ㅠㅡㅡㅜ... 이거랑 annotate로 만든것들도 prefetch 해주고싶지만.. 쉽지않네요......

soheon-lee commented:
화이링
